### PR TITLE
Support auto_parallel

### PR DESCRIPTION
**Pull request by wyg1997**
使用 oneflow 仓库 feat-auto_parallel 分支，通过配置 `graph.auto_parallel=False/True` 可以关闭/开启自动并行。

及 bert 为例，执行以下命令即可：

```shell
COST_RATIO=0.05 WAIT_TIME=1.65e4 AUTO_PARALLEL_TRANSFER_COST=1.65e4 \
    ./tools/train.sh tools/train_net.py configs/bert_large_pretrain.py 4 graph.auto_parallel=True
```

### CODE REVIEW

File: libai/engine/trainer.py [hunk 1]
```diff
@@ -313,8 +313,11 @@ def run_step(self, get_batch: Callable):
         start = time.perf_counter()
 
         # If you want to do something with the data, you can wrap the dataloader.
-        data = next(self._data_loader_iter)
-        data = get_batch(data, getattr(self.data_loader, "mixup_func", None))
+        if not hasattr(self, "_data"):
+            data = next(self._data_loader_iter)
+            data = get_batch(data, getattr(self.data_loader, "mixup_func", None))
+            self._data = data
+        data = self._data
```
Comment (wyg1997) [inline libai/engine/trainer.py:320]: 这部分使用固定数据，除掉 dataloader 的影响

File: libai/models/utils/graph_base.py [hunk 2]
```diff
@@ -79,12 +81,38 @@ def __init__(
         self.config.allow_fuse_cast_scale(True)
 
         # dist_util = dist.get_dist_util()
+        # auto_parallel
+        self.config.enable_auto_parallel(auto_parallel)
+        import os
+        cost_ratio = float(os.environ["COST_RATIO"])
+        wait_time = float(os.environ["WAIT_TIME"])
+        trans_time = float(os.environ["AUTO_PARALLEL_TRANSFER_COST"])
```
Comment (wyg1997) [inline libai/models/utils/graph_base.py:89]: 自动并行的几个参数用环境变量来控制，跑其它模型的时候先用 PR 描述中的参数就行，后面有效果不好了可以再调

File: tools/train_net.py [hunk 3]
```diff
@@ -13,6 +13,11 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import oneflow as flow
+
+flow.boxing.nccl.enable_use_compute_stream(False)
+flow.boxing.nccl.enable_all_to_all(True)
```
Comment (wyg1997) [inline tools/train_net.py:19]: - 1d sbp 时不使用 compute_stream，使计算和传输可以更好重叠
- P->B 走 nccl 而不是 slice boxing

**Comment by thinksoso** [comment]
# 自动并行测试
## bert
### 开启自动并行
[05/19 16:12:15] lb.utils.events INFO:  eta: 0:00:00  iteration: 9999/10000  consumed samples: 320000  total_loss: 0.0001473  time: 0.0794 s/iter  data_time: 0.0000 s/iter total throughput: 403.12 samples/s lr: 1.00e-06  
[05/19 16:12:15] lb.engine.hooks INFO: Overall training speed: 9998 iterations in 0:13:13 (0.0794 s / it)
[05/19 16:12:15] lb.engine.hooks INFO: Total training time: 0:13:31 (0:00:17 on hooks)
[05/19 16:12:15] lb.engine.hooks WARNING: Given val metric Acc@1 does not seem to be computed/stored. Will not be checkpointed based on that.

## gpt2
### 开启
[05/17 01:26:48] lb.utils.events INFO:  eta: 0:00:00  iteration: 9999/10000  consumed samples: 160000  total_loss: 0.0001356  time: 0.3719 s/iter  data_time: 0.0000 s/iter total throughput: 43.02 samples/s lr: 1.50e-06  
[05/17 01:26:48] lb.engine.hooks INFO: Overall training speed: 9998 iterations in 1:01:58 (0.3720 s / it)
[05/17 01:26:48] lb.engine.hooks INFO: Total training time: 1:02:14 (0:00:15 on hooks)
[05/17 01:26:48] lb.engine.hooks WARNING: Given val metric Acc@1 does not seem to be computed/stored. Will not be checkpointed based on that.

## t5
### 开启
[05/19 16:12:15] lb.utils.events INFO:  eta: 0:00:00  iteration: 9999/10000  consumed samples: 320000  total_loss: 0.0001473  time: 0.0794 s/iter  data_time: 0.0000 s/iter total throughput: 403.12 samples/s lr: 1.00e-06  
[05/19 16:12:15] lb.engine.hooks INFO: Overall training speed: 9998 iterations in 0:13:13 (0.0794 s / it)
[05/19 16:12:15] lb.engine.hooks INFO: Total training time: 0:13:31 (0:00:17 on hooks)
[05/19 16:12:15] lb.engine.hooks WARNING: Given val metric Acc@1 does not seem to be computed/stored. Will not be checkpointed based on that.

## t5 pipleline
### 开启
[05/19 20:02:45] lb.utils.events INFO:  eta: 0:00:00  iteration: 9999/10000  consumed samples: 320000  total_loss: 0.0003757  time: 0.2119 s/iter  data_time: 0.0000 s/iter total throughput: 150.99 samples/s lr: 1.00e-06  
[05/19 20:02:45] lb.engine.hooks INFO: Overall training speed: 9998 iterations in 0:35:19 (0.2120 s / it)
[05/19 20:02:45] lb.engine.hooks INFO: Total training time: 0:35:35 (0:00:16 on hooks)
[05/19 20:02:45] lb.engine.hooks WARNING: Given val metric Acc@1 does not seem to be computed/stored. Will not be checkpointed based on that.

**Pull request closed without merging**